cosa/20200805-meta-merge.md

```diff
@@ -0,0 +1,103 @@
+# Merging of build meta-data
+- Specification ID: cosa-20200804
+- Supersedes: None
+- Conflicts: None
+
+# Background
+
+One of the biggest blockers to performing parallel, diustributed and/or re-starting failed builds is the ability to lock write access to `meta.json` in COSA.
+
+Under the following assmptions, COSA needs to support merging of `meta.json`:
+- Not all `cmd-buildextend-*` operations will happen on the same pod or Linux namespace
+- S3 is the authoritative store for builds. S3 is eventually consistent meaning that the version read from S3 may be replaced after reading from S3. Implementing distributed locking on `meta.json` based on S3 is problematic.
```

Comment:
Correct.  Since the `meta.json` is only ever written twice, the race for eventual consistency is mitigated. I envision that in distributed mode the orchestrator would despot the artifact into S3 itself with `meta.json.<ARTIFACT>`.  Since `cmd-upload-build` won't overwrite the files that exist. So the flow would be:
- create base build and run `cmd-upload-build`
- distributed builds for $ARTIFACT and call `cmd-upload-build` at the end of each
- sync `meta.json.<ARTIFACT>`  and run `cmd-meta-merge` and upload final `meta.json`

Reply:
Sounds good. It might be worth adding just a little more detail in here to spell that out. It's subtle, but important.